[3.x] Fix navigation merge errors on too small triangles

Scony opened the pull request:
Fixes #56786 and part of #57254

See https://github.com/godotengine/godot/issues/56786#issuecomment-1014929078 for detailed explanation.

Please note this change may potentially have a severe impact on navigation, so it would be nice if someone else can test the lack of errors as well as Agents traversing navmesh.

Once this is accepted I'll open similar for `master`.

@Duroxxigar @AndreaCatania

akien-mga commented on modules/navigation/nav_map.cpp:
```diff
@@ -606,6 +606,11 @@ void NavMap::sync() {
 		for (size_t poly_id(0); poly_id < polygons.size(); poly_id++) {
 			gd::Polygon &poly(polygons[poly_id]);
 
+			// skip triangles which are not triangles after aliging vertices towards cells (see issue #56786)
```
Our preferred comment style is to use complete sentence formatting (also typo fix, and change to make issue references easier to grep):
```suggestion
			// Skip triangles which are not triangles after aligning vertices towards cells (see GH-56786).
```

Scony replied:
done, thanks for suggestion!

Scony commented:
I've created a proper demo to visualize paths going through "small" triangles which I've skipped and it turns out the triangles are not bypassed. They're removed. See the demo with drawing random paths on generated navigation with 2 merge errors:
![2022-01-18-195711_4480x1440_scrot](https://user-images.githubusercontent.com/1207385/150001665-30f69613-91aa-47db-8399-c9ff8ebd472a.png)

And hence I need to improve it further so they're taken into account.

Scony commented:
While analyzing further I've found yet another problem - when calculating cell key (id) we're dividing `Y` coordinate by `cell_size` while `cell_height` should be used.

With additional precision improvement, everything works as expected, but it's still not perfect.

I'll sync regarding it with @AndreaCatania offline.

Scony committed:
Fix navigation merge errors, fixes #56786

- improved `detail/sample_max_error` default value
- improved floating point precision handling in cell key calculations
- improved `merge error` error message
- exposed `cell_height` of `nav_map` to the `Navigation`
- fixed cell key `y` calculation

Scony commented:
I had to change `detail/sample_max_error` default to `5` so that vertices computed by recast are aligned with the voxels grid much better. This way we're avoiding problems even on very complex meshes while sacrificing some navmesh details.

In practice (flat meshes or meshes with some gentle hills) this makes no difference while this way we're able to avoid potential problems.

The grid alignment algorithm will have to be re-done, but IMO this commit at least polishes what's already there.

AndreaCatania commented on modules/navigation/nav_map.cpp:
```diff
@@ -61,15 +62,23 @@ void NavMap::set_cell_size(float p_cell_size) {
+	const real_t raw_cell_key_z = p_pos.z / cell_size;
+	const int x = int(Math::is_equal_approx(raw_cell_key_x, Math::ceil(raw_cell_key_x)) ? Math::ceil(raw_cell_key_x) : Math::floor(raw_cell_key_x));
+	const int y = int(Math::is_equal_approx(raw_cell_key_y, Math::ceil(raw_cell_key_y)) ? Math::ceil(raw_cell_key_y) : Math::floor(raw_cell_key_y));
+	const int z = int(Math::is_equal_approx(raw_cell_key_z, Math::ceil(raw_cell_key_z)) ? Math::ceil(raw_cell_key_z) : Math::floor(raw_cell_key_z));
```
With this approach it uses the `Math::ceil(raw_cell_key_...)` value only if the `raw_cell_key_...` is near to that value, which make this algorithm not that different from the original one: for example if the value is `0.9` it will still use `0` (floor) as value, since `0.9` is not enough near to `1.0` (ceil).

I wonder if the following one is better, considering that it returns always the closest integer:
```c++
	const int x = static_cast<int>(Math::round(p_pos.x / cell_size));
	const int y = static_cast<int>(Math::round(p_pos.y / cell_size));
	const int z = static_cast<int>(Math::round(p_pos.z / cell_size));
```

If the value is `0.9` it will use `1.0` as round returns the nearest integral. What do you think?

Scony replied:
Well, haven't thought about it but now that I think of it, it actually should work and the code will be much cleaner. I'll give it a try.

AndreaCatania replied:
Thinking more about it I just realized that my above proposed solution will likely change how the algorithm is supposed to work, and that will cause some issues. In any case I still think we need a much simpler way to obtain that `PointKey`. The following one is a similar solution to the one you proposed, what do you think? 
```cpp
const int x = static_cast<int>((p_pos.x / cell_size) + CMP_EPSILON);
const int y = static_cast<int>((p_pos.y / cell_size) + CMP_EPSILON);
const int z = static_cast<int>((p_pos.z / cell_size) + CMP_EPSILON);
```

Scony replied:
@AndreaCatania I've tested the solution with `Math::round` and it's both - fine (tested on various cases) and clean (easy to read and understand + it properly handles precision) and thus I think that will do.

AndreaCatania reviewed (commented):
It looks nice! Great job, just a small comment on the function `get_point_key`.

AndreaCatania reviewed (approved):
Looks good!! Thanks 🍉

akien-mga commented:
Thanks!
